src/reducers/app.js

```diff
@@ -66,17 +75,219 @@ function hasZoomedViaMousewheel(state: boolean = false, action: Action) {
   }
 }
 
+/**
+ * This function ensures that the state transitions are logical and make sense. The
+ * switch statement provides a mapping of what states are valid to transition to from
+ * a previous state. A switch is used rather than an object in order to make it easier
+ * to exhaustively check with Flow.
+ */
+function _validateStateTransition(
+  prev: ZipFileState,
+  next: ZipFileState
+): ZipFileState {
+  const prevPhase = prev.phase;
+  let expectedNextPhases;
+  switch (prevPhase) {
+    case 'NO_ZIP_FILE':
+      expectedNextPhases = [
+        // Coming into a fresh page load with a zip file, this is the first state
+        // transition.
+        'LIST_FILES_IN_ZIP_FILE',
+        // This will happen if coming in from a URL that is already pointing at a file
+        // in a zip.
+        'PROCESS_PROFILE_FROM_ZIP_FILE',
+      ];
+      break;
+    case 'LIST_FILES_IN_ZIP_FILE':
+      expectedNextPhases = [
+        'PROCESS_PROFILE_FROM_ZIP_FILE',
+        'FILE_NOT_FOUND_IN_ZIP_FILE',
+      ];
+      break;
+    case 'PROCESS_PROFILE_FROM_ZIP_FILE':
+      expectedNextPhases = [
+        'VIEW_PROFILE_IN_ZIP_FILE',
+        'FAILED_TO_PROCESS_PROFILE_FROM_ZIP_FILE',
+        // When navigating with the URL, it's possible to go back and list the files.
+        'LIST_FILES_IN_ZIP_FILE',
+      ];
+      break;
+    case 'FILE_NOT_FOUND_IN_ZIP_FILE':
+      expectedNextPhases = ['LIST_FILES_IN_ZIP_FILE'];
+      break;
+    case 'FAILED_TO_PROCESS_PROFILE_FROM_ZIP_FILE':
+      expectedNextPhases = ['LIST_FILES_IN_ZIP_FILE'];
+      break;
+    case 'VIEW_PROFILE_IN_ZIP_FILE':
+      expectedNextPhases = ['LIST_FILES_IN_ZIP_FILE'];
+      break;
+    default:
+      throw new Error(`Unhandled ZipFileState “${(prevPhase: empty)}”`);
+  }
```

Comment:
Mainly I was trying to find a way to use finite state machines that can describe how it can go from one state to the next. I found it very useful for handling every case when I was writing the code. It helped me find a couple of surprising state transitions when I was working on the code. I do agree that it adds a bit of complexity.

I guess my preference would be to keep it in, as I feel like it helps reason with some complex async behavior. It does increase code complexity. If the error throwing scares you, I could put in a check that we are in development mode.

Reply:
As far as rewriting the switch, I reasoned with all the code that way, so I'd prefer not to switch it up (pun intended). I'm not seeing the benefit of rewriting it.

Reply:
> I found it very useful for handling every case when I was writing the code. It helped me find a couple of surprising state transitions when I was working on the code. 

I wonder if this isn't what tests should be for, instead of application code ?